kong/plugins/oauth2/access.lua

```diff
@@ -450,19 +450,18 @@ end
 
 local function parse_access_token(conf)
   local found_in = {}
-  local result = retrieve_parameters()["access_token"]
-  if not result then
-    local authorization = ngx.req.get_headers()[conf.auth_header_name]
-    if authorization then
-      local parts = {}
-      for v in authorization:gmatch("%S+") do -- Split by space
-        table.insert(parts, v)
-      end
-      if #parts == 2 and (parts[1]:lower() == "token" or parts[1]:lower() == "bearer") then
-        result = parts[2]
-        found_in.authorization_header = true
-      end
+  local access_token = ngx.req.get_headers()[conf.auth_header_name]
```

Comment:
While you are at it, could you also fix issue with this code where it actually crashes runtime, if you happen to send the header more than once. Making `local access_token` be `table` instead of string. What to do in such case, well just pick the first for example `access_token[1]`.

Reply:
Crashed when you tested? I am not able to reproduce such a failure, do I have to run in debug mode to see it?

Reply:
I think on line 456:
```lua
 for v in access_token:gmatch("%S+") do
```

If the access token is actually a table it will crash, as `table`s in Lua don't have `gmatch`. It is something that is only in `string`.

As you see here:
https://github.com/openresty/lua-nginx-module#ngxreqget_headers

The example where it sends headers:
```http
 Foo: foo
 Foo: bar
 Foo: baz
```

The `ngx.req.get_headers()["Foo"] ` will return a table instead of string containing one of those values:
```lua
 {"foo", "bar", "baz"}
```

There is other way, which is `ngx.var.http_Foo` that always returns a single value, but I think using `ngx.req.get_headers()` there is fine.

Reply:
What I mean is to add code like this:
```
if type(access_token) == "table" then
  access_token = access_token[1]
end
```

To make code more bullet-proof.

Reply:
You prefer the above extra check rather than forcing it to 1(requires no extra conditional/assignment?:
access_token = ngx.req.get_headers(1)["Foo"] ? 

I am going to retest again too in debug because I never saw that failure print out to stdout with the lua code, I would have expected to see something of a failure on the gmatch like you mentioned.

Reply:
`max_headers` parameter is not about how many `Foo` headers you get. It is just how many headers you will read in total.

Reply:
Oh, right duh that makes more sense. I just wanted to get fancy with it :P , testing again now just so I can produce the failure too then will go with the table check suggestion.

Reply:
Yeah tested my code again with debug enabled too, I see this in logs when I run through it:

```
2018/09/20 19:58:22 [info] 33#0: *1692 client sent duplicate header line: "Authorization: Bearer SHj7zxYf7mhcyKAb5uLG1A1FtzbtG93Y", previous value: "Authorization: Bearer SHj7zxYf7mhcyKAb5uLG1A1FtzbtG93Y" while reading client request headers, client: 10.xxx.xx.x, server: kong, request: "POST /api/fasthttp/v1.0 HTTP/1.1"
```

But that was it for logs so must be a silent failure to not make it to docker std out, regardless I am fine with just taking the first and continuing the call. Thx.

Reply:
@bungle Mr. Bungle sir I believe I have somehow managed to fix the PR with the changes and maintain one commit with a fixed commit message as per Contribution guidelinse as requested by @Tieske earlier using a combination of git commit --amend's and git push -f . Let me know if anything else is off. And I think you are spot on with that evaluation, I was leaving it to the default and nginx must be handling that gracefully before lua land was seeing it.

Thanks!

Reply:
@jeremyjpj0916 Appreciate the effort :) The proper prefix however is `perf(oauth2)` (as per the [contributing docs](https://github.com/Kong/kong/blob/master/CONTRIBUTING.md#commit-message-format): `<type>(<scope>)`. We also enforce a non-capitalization of the git message headers :)

Ideally, I believe that a clearer message for the intent behind this commit would be something similar to:
```
perf(oauth2) do not read request body if unnecessary

`access_token` retrieval used to unconditionally read the request body.
We now evaluate if the token is contained in the request headers before
attempting to read the body.

Fix ...
```

This way, the commit message is roughly organized as:
```
<type>(<scope>) <intent>

<before patch>
<new behavior>

<links>
```

Reply:
I think I modified everything to your request 👍  . It may not be today, or tomorrow, or within the month. But one day I hope I can aspire to get it right well enough that my time spent editing the commit messages and PR takes less time that the code itself lol.

Reply:
@jeremyjpj0916 Thank you! I do spend time editing the commit messages/style issues myself when I find some in PRs, but:

1. I am a fervent believer that contributors should learn how to "fit" into a codebase/process and distillate the knowledge and practices themselves to future newcomers :)
2. If you are to contribute more and more frequently, "teach a man how to fish..." yada yada, right? :)